### ISSUE TITLE
Higher level (`struct` based) process abstraction

### ISSUE THREAD

**Issue by bkolobara** [lunatic-solutions/rust-lib#10]
Defining and spawning processes are fundamental tasks you do when working with lunatic. Naturally, we want to make the developer experience around them as pleasant as possible. I would like to introduce a new higher-level Rust API for defining processes that is easier to use, provides powerful abstractions (like hot-reloading) and works nicely with Rust's type system. Before going into details of this proposal I would like to take a step back just to explain how the current system works and how we got there.

Lunatic allows you to [point to a function](https://docs.rs/lunatic/0.7.0/lunatic/process/fn.spawn.html) in your Rust application and spawn a process from it. This is how Erlang/Elixir works too; and it's a really simple yet powerful tool.

However, lunatic is at the same time a "system" to run any WebAssembly module as a process. This means that it can dynamically load `.wasm` files written in different languages, [spawn processes](https://docs.rs/lunatic/0.7.0/lunatic/struct.Module.html#method.spawn) and send messages to them. Contrary to spawning a process from the currently running module, we can't have any type system guarantees about what messages the spawned process can receive. We may not even have access to the source code of the loaded WebAssembly module.

In the rest of this text I'm solely going to focus on the act of spawning processes from the currently running module where we actually can utilise Rust's type system.

### History

From the first days of using Rust with lunatic, I have envisioned that you can just simply spawn processes from functions, the same way you can do it in Erlang/Elixir. A big obstacle here is the fact that Rust is a strongly typed language and Erlang/Elixir are dynamic. How do you fit a concept of processes and messages into the type system? How could you send different types of messages to a process and have the type system catch errors during compilation?

My [first approach](https://docs.rs/lunatic/0.3.10/lunatic/channel/index.html) just mimicked the **channels** approach that is used in other Rust libraries and in [Go](https://gobyexample.com/channels). This means that the message type was bound to the channel and a process could capture many channels on startup. This is not ideal, as it moves away from the single mailbox principle and suddenly you have as many "mailboxes" as you have channels. Also, you can't simultaneously wait on multiple channels, as their return values may be of different type. You would always end up capturing one channel and wrapping all the different message types in a super-type `enum`. This resulted in me getting rid of the channels approach and just having a [single mailbox](https://docs.rs/lunatic/0.7.0/lunatic/struct.Mailbox.html) that is received as an argument of the process entry function. This is basically what we have today.

I'm generally satisfied with the current approach and believe that it gives you a simple way to spawn one-off processes and compile time errors if you try to send messages of wrong types to a process (that can't handle them).

### Proposal

This "function as a process" approach starts to fall apart once you have more complex processes with complicated behaviours. For example, you can't force a process to return you a message. There is no type level support with the current system to enforce that if a process receives one type of message it should respond with a specific type. Ideally we would like to be able to express this kinds of contracts with the Rust type system.

Erlang/Elixir has a higher level abstraction, the [GenServer](https://hexdocs.pm/elixir/1.12/GenServer.html) (generic server). You can implement two kinds of behaviours:

* handle_call - You received a request and are supposed to respond with a reply.
* handle_cast - Another process sent you a message, but is not waiting for a response.

I would like to bring the same concept to Rust's lunatic library.

#### Example

Another library in the Rust ecosystem already figured out a good approach on modelling such behaviours inside Rust's type system, [Actix](https://github.com/actix/actix). If we represent a process state as a `struct` we can define different message handlers on it. The new API would look something like this:

```rust
// A message
#[derive(Serialize, Deserialize)]
struct Sum(usize, usize);

// The process state
struct Calculator;

impl lunatic::Process for Calculator {
    type Context = Context<Self>;
}

// A handler for `Sum` messages
impl HandleCall<Sum> for Calculator {
    type Result = usize; // <- response type

    fn handle(&mut self, msg: Sum, ctx: &mut Context<Self>) -> Self::Result {
        msg.0 + msg.1
    }
}

fn main() {
    let addr = Calculator.start();
    let res = addr.send(Sum(10, 5));

    match res {
        Ok(result) => println!("SUM: {}", result),
        _ => println!("Communication to the actor has failed"),
    }
}
```

In this example we get compile time guarantees that all the messages sent & received are of the correct type. We also can force the process to respond with an appropriate type (`Self::Result`) when a message of a specific type (`Sum`) is sent. If we left the response out, the code would not even compile.

#### Hot reloading

The same way Erlang's [GenServer](https://hexdocs.pm/elixir/1.12/GenServer.html) makes it easier to do hot-reloading of Erlang processes, once we have more structure (pun intended) around the processes we can also introduce process lifecycles that make it possible to accomplish hot reloading. We can enforce that the process state implements `Serialize + Deserialize` and on code changes we just serialize the process state and deserialize it as part of the new implementation.

The `Process` trait could provide default implementations if the state structure stayed the same, but should also allow developers to define "state transitions" to new versions:

```rust
enum Action {
    Reset,    // The process is re-spawned with a new state.
    HotUpdate // Hot-reload the process and try to reuse the previous state. 
}

impl lunatic::Process for State {
    type Context = Context<Self>;

    fn update_behaviour() -> Action { Action::HotUpdate }
    fn update(old_state: Data) -> Result<Self, UpdateError> {
        // old_state contains a serialized version of the previous `Self`.
        // The implementation of this method needs to deserialize it and create a new
        // version of `Self`.
    }
}
```

This would also make it possible to move processes between machines. If the state can be serialized it can be moved to another node and a process could be bootstrapped there from it.

All of these features don't require any changes to the lunatic runtime and can be completely implemented as a library on the currently existing primitives. I think that this is an important characteristic of lunatic, we can keep the underlaying runtime lean, simple and performant, but build really powerful abstractions on top of it.

### Summary

I believe that by adding this kind of API we can lean much harder on Rust's type system to enforce correctness. At the same time it nicely mimics Erlang's proven approach.

This is not meant to replace the function based API, it's more of an augmentation. It represents a philosophy on how to structure your state, requests, responses and handle lifecycle (code updated). The function based API still gives you full flexibility to programatically handle messages and is really convenient when creating processes from small closures (timeouts, etc.).

**Comment by sleipnir** [comment]
In Actix this is done through System as:

```rust
fn main() {
     let mut system = System::new();

     let addr = system.block_on(async { Calculator.start() });

     system.run();
}
```

In this case it also uses a start function, but this is passed to System, I don't like this Actix approach but as it is done through System I consider it ok.
This is also good for location transparency purposes as you only get a reference to the Process so it doesn't matter if the Process is local or remote and this is made explicit in the API

**Comment by jvdwrf** [comment]
Yes, I would agree with your comments. Right now the API looks as following (Naming is temporary):
```rust
fn main() {

    let my_actor = MyActor { count: 0 };
    let gen_server = GenServer::new(my_actor);
    let process = gen_server.start();

    process.cast("hi".to_owned());
    process.cast(32);

    let _reply = process.call(32).unwrap();
}

#[derive(Serialize, Deserialize)]
struct MyActor {
    count: u32,
}

impl Handle<u32> for MyActor {
    type Reply = u32;

    fn handle_cast(&mut self, msg: u32) {
        self.count += msg;
    }

    fn handle_call(&mut self, msg: u32) -> Self::Reply {
        self.count += msg;
        self.count
    }
}

impl Handle<String> for MyActor {
    type Reply = String;

    fn handle_cast(&mut self, _msg: String) {}

    fn handle_call(&mut self, msg: String) -> Self::Reply {
        msg
    }
}
```

I think this would work a lot better, and in the future, other GenServer implementations can be created easily.

**Comment by jvdwrf** [comment]
1. I have thought about this, but am not sure whether this would make the api any better, right now I want to have the GenServer as a struct, but implement a spawn trait for any actors. Then they get a context where handle functions can be called, this is what internally the GenServer will use.
I think I will try my approach first, changing the api afterward is not so much work I think.

2.  Yes I will do this, but for right now this is simpler for testing some stuff out

**Comment by bkolobara** [comment]
Closed by #15!
